docs: refactor kratos & update docs on SMTP config and email templates (#946)
ory/docs: docs/kratos/emails-sms/05_custom-email-templates.mdx

---
id: custom-email-templates
title: Customize message templates for recovery and verification emails
sidebar_label: Email templates

> ---

Author: @Benehiko I'd appreciate it if you could review this document :)

Reviewer: Hey @tomekpapiernik 
I just tried out the approach with Ory Cloud and it worked fine :) 
The nested templates section needs an Ory Cloud part too right?

Reviewer: Oh misread that, so "Nested templates containing remote templates" are possible only in self-hosted right now and normal nested should work fine? in that case I can do some testing.

Author: "Normal" as in all the logic has to be in a single template? Yes. 
Move the "nested with remote" thing to that self-hosted doc?

Ory Identity Service (Ory Kratos) comes with built-in templates for all messages sent by the system. You can replace the default
templates with custom ones that will carry your own branding, visual elements, and communication tone.

The system uses templates in these flows:

1. `recovery` - the system sends these messages when users recover their accounts, for example after forgetting their password.
   - `recovery.valid` - the system sends these messages when the user completes the recovery flow successfully.
   - `recovery.invalid` - the system sends these messages when the recovery flow fails.
2. `verification` - the system sends these messages when users verify their accounts after signing up.
   - `verification.valid` - the system sends these messages when the user verifies their account successfully.

>    - `verification.invalid` - the system sends these messages when the account verification fails.

Reviewer: ```suggestion
   - `verification.invalid` - the system sends these messages when the account verification fails for example when the account does not exist.
```
```suggestion
   - `verification.invalid` - the system sends these messages when the user tries to verify an account that does not exist.
```

Maybe we can make it clearer when these messages are sent?

Author: Sure! Did that for `recovery.invalid`, but your example doesn't really sound right... Are you sure you can try to verify an email that's not "registered"? I don't think that's possible, but maybe I'm missing something.

Author: Ok, discussed this with Alano and added examples for both invalid "flows".

<Tabs>
  <TabItem value="cloud" label="Ory Cloud" default>

To use custom templates in Ory Cloud, encode the content template with Base64 and use the encoded string to update the Identity Service config.
Follow these steps:

1. List all your projects and get the Identity Service config file:

   ```shell
   ## List all available projects
   ory list projects

   ## Get config
   ory get identity-config <project-id> --format yaml > identity-config.yaml
   ```
2. Add your Base64-encoded custom message templates to the config:

   ```yaml title="identity-config.yaml"
   courier:
     smtp:
       from_name: MyProject via Ory
     templates:
       recovery:
         invalid:
           email:
             body: {
               "html": "base64://ENCODED_HTML_TEMPLATE",
               "plaintext": "base64://ENCODED_PLAINTEXT_TEMPLATE"
             }
         valid:
           email:
             body: {}
       verification:
         invalid:
           email:
             body: {}
         valid:
           email:
             body: {}
    ```
3. Upload the edited config:

   ```shell
   ory update identity-config <project-id> --file updated_config.yaml
   ```

:::warning

> In Ory Cloud, you can use only Base64-encoded templates.

Author: I change the message to strengthen the point that b64 URIs are the only way to go. Added an explanation of what happens when you try to provide templates using unsupported methods (fallback to default templates). 

 Reg. `template_override_path` - I describe it only in the "self-hosted section" for a reason. ;) It's not a sustainable approach to say that certain things from self-hosted don't work in Cloud and vice-versa. Instead, the docs draw the line by explicitly saying to what scenario a given piece of instructions applies. In this document, it's done using the tabs. Fun fact - when the user switches one tab to, let's say, "self-hosted", all other tabs on the page switch to self-hosted as well.

> In Ory Cloud, you can use only Base64-encoded templates.

Reviewer 2: True, but since we don't offer some of these "features" in Cloud, people might automatically think self-host and cloud should be 1:1 on "features" but the way we configure them might be different.

I would just strengthen on key areas that it is not possible in cloud in the cloud section (can even be a warning or note at the top of the cloud section).

" in Cloud setting `template_override_path` is unsupported as well as setting any remote URL for template `body` and `subject`..."

      valid:
        email:
          body:
            html: https://some-remote-resource/gotmpl
            plaintext: base64://SGV5IHlvdSBkZWNvZGVkIG1lIDop
          # optional
          subject: file://some-file/subject.gotmpl
```

:::info

When self-hosting Ory Kratos, the system expects the templates at these paths:

- `<kratos-root>/<template-root>/recovery/valid`
- `<kratos-root>/<template-root>/recovery/invalid`
- `<kratos-root>/<template-root>/verification/valid`
- `<kratos-root>/<template-root>/verification/invalid`

Read [this document](../self-hosted/mail-courier-selfhosted) to learn more about the email courier and the specifics of using
templates in self-hosted Ory Kratos instances.

:::

  </TabItem>
</Tabs>
````

## Creating templates

Templates use the golang template engine in the `text/template` package for rendering the `email.subject.gotmpl` and
`email.body.plaintext.gotmpl` templates, and the `html/template` package for rendering the `email.body.gotmpl` template.

Learn more:

- https://pkg.go.dev/text/template
- https://pkg.go.dev/html/template

:::tip

Templates can use the [Sprig](https://github.com/Masterminds/sprig) library, which provides

> [more than 100 commonly used template functions](http://masterminds.github.io/sprig/)

Author: Added this list as a warning, TY

:::

Use these variables in templates:

- `recovery.valid` - `To`,`RecoveryURL`, and `Identity` for validating recovery.
- `recovery.invalid` - `To` for invalidating recovery.
- `verification.valid` - `To`, `VerificationURL` and `Identity` for validating verification.
- `verification.invalid` - `To` for invalidating verification.

For example:

[`https://github.com/ory/kratos/blob/master/courier/template/courier/builtin/templates/verification/valid/email.body.gotmpl`](https://github.com/ory/kratos/blob/master/courier/template/courier/builtin/templates/verification/valid/email.body.gotmpl)

```gotmpl title="courier/template/templates/verification/valid/email.body.gotmpl"
Hi, please verify your account by clicking the following link:

<a href="{{ .VerificationURL }}">{{ .VerificationURL }}</a>
```

```gotmp title="courier/template/templates/verification/valid/email.body.plaintext.gotmpl"
Hi, please verify your account by clicking the following link: {{ .VerificationURL }}
```

> 

Author: Done

## Customizing template content for specific users

To enable customizing the content of templates based on the identity of the recipient of the email, the Ory identity is available
as the `Identity` object. This object is a map containing all the attributes of an identity, such as `id`, `state`,

> `recovery_addresses`, `verifiable_addresses` and `traits`.

Author: I've covered this in the i18n example :) Added some more meat to that section, PTAL.

content changes if the system detects `fr` or `it` in the identity traits. If the system detects any other language, it uses the `eng` template.

```gotmpl
{{define "en"}}
Hi,

Please recover access to your account by clicking the following link:

{{ .RecoveryURL }}
{{end}}

{{define "fr"}}
Bonjour ,

Veuillez récupérer l'accès à votre compte en cliquant sur le lien suivant :

{{ .RecoveryURL }}
{{end}}

{{define "it"}}
Ciao ,

Recuperare l'accesso al proprio account facendo clic sul seguente link:

{{ .RecoveryURL }}
{{end}}

{{- else if eq .Identity.traits.language "fr" -}}
{{ template "fr" . }}
{{- else if eq .Identity.traits.language "it" -}}
{{ template "it" . }}
{{- else -}}
{{ template "en" . }}
{{- end -}}
```

  </TabItem>
  <TabItem value="selfhosted" label="Self-Hosted Ory Kratos Config" default>

To create nested templates, you can use either in-line template definitions, or as in this example, use separate templates. The following
example defines the message body for verification emails. It assumes that there is a `lang` attribute defined in the identity traits
that contains the preferred language of the user.

> 

Reviewer 2: I think also add a warning here about using separate templates in Ory Cloud. They aren't supported since you need to specify the `template_override_path` and this cannot be specified in Cloud.

Author: Same comment as in the `template_override_path` case. Additionally, `template_override_path` is not in the Cloud config by default, you'd have to go out of your way to add it manually.

Reviewer 2: Well people got confused in our Slack channel on what is possible in Cloud due to what is possible in Open Source. I think just strengthening what you can and cannot do is always good since it leads to less repetitive questions in slack and GH. 

As I mentioned in a previous comment, maybe adding what isn't supported in Cloud at the top of this section might then be best, since you aren't repeating yourself across all of the docs and it is in one place for these particular differences.

```txt title="<template-root>/recovery/valid/email.body.gotmpl"
{{- if eq .Identity.traits.language "de" -}}
{{ template "email.body.de.gotmpl" . }}
{{- else -}}
{{ template "email.body.en.gotmpl" . }}
{{- end -}}
<a href="{{ .RecoveryURL }}">{{.RecoveryURL }}</a>
```

```txt title="<template-root>/recovery/valid/email.body.de.gotmpl"
Hallo {{ upper .Identity.traits.firstName }},

Um Ihr Konto wiederherzustellen, klicken Sie bitte auf den folgenden Link:
```

```txt title="<template-root>/recovery/valid/email.body.en.gotmpl"
Hello {{ upper .Identity.traits.firstName }},

to recover your account, please click on the link below:
```

The i18n template is nested in the root `email.body.gotmpl` template. The sub-templates must conform to the `email.body*` pattern.
The identity of the user is available in all templates.

  </TabItem>
</Tabs>
````

:::tip

> You can use Sprig functions in the nested templates.

Reviewer 2: Reference back to the Sprig section

:::

### Nested templates containing remote templates

:::warning

This approach is available in self-hosted Ory Kratos instances only. For security reasons, Ory Cloud doesn't allow linking to
external resources. As such, all options must be contained in a single template.

:::

When remote templates are used in self-hosted Ory Kratos, the dynamics of loading nested templates change. Remote templates can't

> reference templates outside of the instance's template directory.

Reviewer 2: I don't think this makes sense. In Ory Kratos we need to use a nested template (as you have in the previous section) **when using remote templates**. So this is about using nested remote templates.

We can actually remove a lot of this section and move it to the above section with warnings. Ory Cloud **can** use remote templates and it **can** use nested templates, but when loading them in it has to be base64 encoded.

Author: I think I'm gonna need to jump on a call with you and learn how this nesting remote stuff mechanism works because atm I obviously don't 😅.

Reviewer 2: sure thing :)

smtp:
  template_override_path: /conf/courier-template
  # ...
  templates:
    verification:
      valid:
        email:
          body:
            html: https://some-remote-resource/gotmpl
            plaintext: base64://SGV5IHlvdSBkZWNvZGVkIG1lIDop
          # optional
          subject: file://some-file/subject.gotmpl
```

:::info

When self-hosting Ory Kratos, the system expects the templates at these paths:

- `<kratos-root>/<template-root>/recovery/valid`
- `<kratos-root>/<template-root>/recovery/invalid`
- `<kratos-root>/<template-root>/verification/valid`
- `<kratos-root>/<template-root>/verification/invalid`

Read [this document](../self-hosted/mail-courier-selfhosted) to learn more about the email courier and the specifics of using
templates in self-hosted Ory Kratos instances.

:::

  </TabItem>
</Tabs>
````

## Creating templates

Templates use the golang template engine in the `text/template` package for rendering the `email.subject.gotmpl` and
`email.body.plaintext.gotmpl` templates, and the `html/template` package for rendering the `email.body.gotmpl` template.

Learn more:

- https://pkg.go.dev/text/template
- https://pkg.go.dev/html/template

> <a name="spring-disabled"></a>

Reviewer 2: does this do anything?

Author: Yes! It's an anchor that takes you from one Spring "tip" to another. It's placed in this line to ensure that the right content gets put in the center of the screen when someone uses it.

To enable customizing the content of templates based on the identity of the recipient of the email, the Ory identity is available
as the `Identity` object. This object is a map containing all the attributes of an identity defined in the identity schema, such
as `id`, `state`, `recovery_addresses`, `verifiable_addresses` and `traits`.

:::tip

Read [this document](../manage-identities/overview) to learn more about the Ory identity and the identity schema.

:::

### Nested templates: i18n customization

You can use nested templates to render `email.subject.gotmpl`, `email.body.gotmpl` and `email.body.plaintext.gotmpl` templates.

To enable i18n customization of templates, customize the identity schema to include the user's preferred communication language.
For example:

```json title="Sample custom Identity Schema with user language"
{
  // ...
  "properties": {
    "traits": {
      "type": "object",
      "properties": {
        // ...
        // highlight-start
        "lang": {
          "type": "string",
          "title": "Your preferred language"
      },
      // highlight-end
      "required": [
        "email"
      ],
      "additionalProperties": false,
    }
  }
}

> ```

Reviewer 2: These can also be added as "meta" attributes so then you don't need to rely on the user supplying the language attribute, but could for instance do this based on location or system language.
https://www.ory.sh/docs/guides/manage-identities/managing-users-identities-metadata

Reviewer 2: just maybe link to meta and say this can be done, no need to provide more examples i think

{{- else -}}
{{ template "email.body.en.gotmpl" . }}
{{- end -}}
<a href="{{ .RecoveryURL }}">{{.RecoveryURL }}</a>
```

```txt title="<template-root>/recovery/valid/email.body.de.gotmpl"
Hallo {{ upper .Identity.traits.firstName }},

Um Ihr Konto wiederherzustellen, klicken Sie bitte auf den folgenden Link:
```

```txt title="<template-root>/recovery/valid/email.body.en.gotmpl"
Hello {{ upper .Identity.traits.firstName }},

to recover your account, please click on the link below:
```

The i18n template is nested in the root `email.body.gotmpl` template. The sub-templates must conform to the `email.body*` pattern.
The identity of the user is available in all templates.

  </TabItem>
</Tabs>
````

:::tip

You can use Sprig functions in the nested templates. For security reasons, some functions are disabled in Ory Cloud.
[Click here to see the list of disabled functions.](#spring-disabled)

:::

### Nested templates containing remote templates

:::warning

This approach is available in self-hosted Ory Kratos instances only. For security reasons, Ory Cloud doesn't allow linking to
external resources. As such, all options must be contained in a single template.

> 

Reviewer 2: nice! :)

}
```

:::tip

> You can get user-specific metadata, such as the language or the locale automatically using the identity metadata. Read

Reviewer 2: ```suggestion
As an administrator, you can also set Identity metadata, such as the language or the locale automatically using the identity metadata property. Read
```

Author: Identity -> identity

thanks!
